### ISSUE TITLE
[no-misused-promises] Not marking as error when combined with a logical operator variable

### ISSUE THREAD

**Issue by LeandroDG** [typescript-eslint/typescript-eslint#2544]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
  rules: {
    // "spaced-comment": ["warn", "always"],
    semi: ["error", "always"],
    quotes: ["warn", "double"],
    "prefer-const": "warn",
    "no-underscore-dangle": "warn",
    "@typescript-eslint/explicit-function-return-type": "warn",
    "@typescript-eslint/no-explicit-any": "off",
    "@typescript-eslint/type-annotation-spacing": "off",
    "@typescript-eslint/no-empty-function": "warn",
    "@typescript-eslint/no-use-before-define": "off",
    "@typescript-eslint/camelcase": "off",
    **"@typescript-eslint/no-misused-promises": "error",**
    "@typescript-eslint/no-floating-promises": "error"
  }
```

```TS
  async function reproduceIssue(parameterConditional: boolean) {
    const myPromise = Promise.resolve(true);
    const localTrue = true;
    const localFalse = false;

    // This correctly shows a bad usage
    if (myPromise) {
      console.log("Bad usage");
    }

    // This correctly shows a bad usage
    if (localTrue && myPromise) {
      console.log("Bad usage");
    }

    // This correctly not marks a bad usage
    if (localFalse && myPromise) {
      console.log("Bad usage");
    }

    // This should show a bad usage
    if (parameterConditional && myPromise) {
      console.log("Bad usage");
    }

    // This should show a bad usage
    if (returnTrue() && myPromise) {
      console.log("Bad usage");
    }
  }

  function returnTrue(): boolean {
    return true;
  }
```

**Expected Result**

These lines should mark the promise as misused, but as they are applied an AND with an unknown boolean parameter or function result, they don't:
```
if (parameterConditional && myPromise) {
if (this.returnTrue() && myPromise) {
```

**Actual Result**

Those lines are not marked as misuses of the promise but they are.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.1.0` |
| `@typescript-eslint/parser`        | `4.1.0` |
| `TypeScript`                       | `3.9.7` |
| `ESLint`                           | `7.8.1` |
| `node`                             | `12.8.3` |

**Comment by bradzacher** [comment]
The rule doesn't do any sophisticated inspection - it just looks for a `Promise` typed thing in a place it shouldn't be.

In this instance.

```ts
    // This correctly not marks a bad usage
    if (localFalse && myPromise) { // typeof === false - hence it doesn't detect it
      console.log("Bad usage");
    }

    // This should show a bad usage
    if (parameterConditional && myPromise) { // typeof = false & Promise<boolean> - hence it doesn't detect it
      console.log("Bad usage");
    }

    // This should show a bad usage
    if (this.returnTrue() && myPromise) { // ditto - false & Promise
      console.log("Bad usage");
    }
```

Happy to accept a PR to make the rule smarter!

**Comment by tadhgmister** [comment]
I don't think this is a matter of "the rule just isn't sophisticated enough", the rule [specifically checks](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/no-misused-promises.ts#L136-L137) that all union parts are then-able:

```ts
// Variation on the thenable check which requires all forms of the type (read:
// alternates in a union) to be thenable. Otherwise, you might be trying to
// check if something is defined or undefined and get caught because one of the
// branches is thenable.
function isAlwaysThenable(checker: ts.TypeChecker, node: ts.Node): boolean {
    ... 
    // If one of the alternates has no then property, it is not thenable in all
    // cases.
    if (thenProp === undefined) {
      return false;
    }
```
I'm not sure what the correct logic should be,  some options are:

1. if any member of the union is then-able in a condition raise a warning, this is easiest but goes against some of the initial design
2. when checking a [logical expression](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/no-misused-promises.ts#L59-L63) conditionally check the right hand side if the expression is directly inside a condition location, will need to handle nesting like `if(a && (b || c)){}` correctly
3. something more complicated like raise a warning if the only other parts of the union are boolean, seems more obscure though.

also minor note, [this](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/no-misused-promises.ts#L64-L68) can be simplified at the same time:
```diff
-      UnaryExpression(node) {
+     'UnaryExpression[operator="!"]'(node: TSESTree.UnaryExpression) {
-        if (node.operator === '!') {
          checkConditional(node.argument);
-        }
      },
```

**Comment by bradzacher** [comment]
The issue here is that the rule does not properly analyse logical expressions within a conditional test context - it's not being sophisticated enough.

It does do _some_ checks on `LogicalExpression`, but it only checks the left expression, never the right. It ignores the right because it makes the assumption that the right value is either:
- being used as per a short-circuit
- handled by the base conditional test

However with this simple logic, it does not work well for all cases. 
For example - how should the `boolean | Promise` type be handled?

Imagine the following code:
```ts
async function foo(arg: false | Promise<boolean>) {
  if (arg) {
    return await arg;
  }
}
```
Is this a misused promise? The answer is no - you need to use the conditional to narrow the type first. So the rule _correctly_ handles this.
(sure you could use `typeof` instead to narrow it - but that shouldn't be required).

However if I do this:
```ts
async function foo(bool: boolean, prom: Promise<boolean>) {
  if (bool && prom) {
    return await prom;
  }
}
```
Is this a misused promise? The answer is yes - it's an unnecessary condition AND it's a misused promise.

The rule right now does not deeply inspect the logical - just the result of the logical.
The return type of the logical here is `false | Promise<boolean>`, which is the exact same type as the first example (which we decided was correctly handled).

One final example to further drive the point home:
```ts
async function foo(bool: true, prom: Promise<boolean>) {
  if (bool || prom) {
    return await prom;
  }
}
```
Is this a misused promise? The answer is yes - again it's an unnecessary condition AND it's a misused promise.
However the return type of the logical here is `true`, which the rule will just ignore as it's not thenable.

----

The rule is not doing any sophisticated checks - it's just doing a naive check of the logical expression. It's checking the _runtime_ value of the logical and ignoring the code itself.

So the fix here is as follows:

In a conditional test context the rule should deeply inspect a LogicalExpression and individually test each and every expression separately.

**Comment by jsone-studios** [comment]
I am working this

**Comment by LeandroDG** [comment]
Thank you so much for your help with this =)
